src/argparse.lua

```diff
@@ -1087,6 +1087,230 @@ function Parser:get_help()
+               :format(table.concat(option._choices, " "))
+         else
+            compreply = 'COMPREPLY=($(compgen -f "$cur"))'
+         end
+         table.insert(opts, (" "):rep(indent + 4) .. pattern)
+         table.insert(opts, (" "):rep(indent + 8) .. compreply)
+         table.insert(opts, (" "):rep(indent + 8) .. "return 0")
+         table.insert(opts, (" "):rep(indent + 8) .. ";;")
+      end
+   end
+
+   if #opts > 0 then
+      table.insert(buf, (" "):rep(indent) .. 'case "$prev" in')
+      table.insert(buf, table.concat(opts, "\n"))
+      table.insert(buf, (" "):rep(indent) .. "esac\n")
+   end
+end
+
+function Parser:_bash_get_cmd(buf)
+   local cmds = {}
+   for idx, command in ipairs(self._commands) do
+      if idx ~= self._help_command_idx then
+         table.insert(cmds, (" "):rep(12) .. ("%s)"):format(table.concat(command._aliases, "|")))
+         table.insert(cmds, (" "):rep(16) .. ('cmd="%s"'):format(command._aliases[1]))
+         table.insert(cmds, (" "):rep(16) .. "break")
+         table.insert(cmds, (" "):rep(16) .. ";;")
+      end
+   end
+
+   if #cmds > 0 then
+      table.insert(buf, (" "):rep(4) .. "for arg in ${COMP_WORDS[@]:1}")
+      table.insert(buf, (" "):rep(4) .. "do")
+      table.insert(buf, (" "):rep(8) .. 'case "$arg" in')
+      table.insert(buf, table.concat(cmds, "\n"))
+      table.insert(buf, (" "):rep(8) .. "esac")
+      table.insert(buf, (" "):rep(4) .. "done\n")
+   end
+end
+
+function Parser:_bash_cmd_completions(buf)
+   local subcmds = {}
+   for idx, command in ipairs(self._commands) do
+      if #command._options > 0 and idx ~= self._help_command_idx then
+         table.insert(subcmds, (" "):rep(8) .. command._aliases[1] .. ")")
+         command:_bash_option_args(subcmds, 12)
+         table.insert(subcmds, (" "):rep(12) .. ('opts="$opts %s"'):format(get_options(command)))
+         table.insert(subcmds, (" "):rep(12) .. ";;")
+      end
+   end
+
+   table.insert(buf, (" "):rep(4) .. 'case "$cmd" in')
+   table.insert(buf, (" "):rep(8) .. self._name .. ")")
+   table.insert(buf, (" "):rep(12) .. ('COMPREPLY=($(compgen -W "%s" -- "$cur"))'):format(get_commands(self)))
+   table.insert(buf, (" "):rep(12) .. ";;")
+   if #subcmds > 0 then
+      table.insert(buf, table.concat(subcmds, "\n"))
+   end
+   table.insert(buf, (" "):rep(4) .. "esac\n")
+end
+
+function Parser:get_bash_complete()
+   local buf = {([[
+_%s() {
+    local cur prev cmd opts arg
+    cur="${COMP_WORDS[COMP_CWORD]}"
+    prev="${COMP_WORDS[COMP_CWORD-1]}"
+    cmd="%s"
+    opts="%s"
+]]):format(self._name, self._name, get_options(self))}
+
+   self:_bash_option_args(buf, 4)
+   self:_bash_get_cmd(buf)
+   if #self._commands > 0 then
+      self:_bash_cmd_completions(buf)
+   end
+
+   table.insert(buf, ([=[
+    if [[ "$cur" = -* ]]
+    then
+        COMPREPLY=($(compgen -W "$opts" -- "$cur"))
+    fi
+}
+
+complete -F _%s -o bashdefault -o default %s]=]):format(self._name, self._name))
```

Comment:
Needs a trailing newline

Reply:
`print` adds a newline
I could add a newline here and use `io.write` if that's preferred.

Reply:
Yeah I think so. I think any user of the `:get_bash_complete()` function would expect a newline to be there.
Of course if you don't want to use `io.write` then two trailing new-lines isn't going to break anything.